src/rdkafka_telemetry_decode.c

```diff
@@ -0,0 +1,536 @@
+                            "Failed to decompress Snappy payload for message "
+                            "(%" PRIusz
+                            " bytes): %s: "
+                            "ignoring message",
+                            inlen, rd_strerror(errno));
+                        rd_free(iov.iov_base);
+                        return -1;  // Indicates decompression error
+                }
+        }
+        *outbuf     = iov.iov_base;
+        *outbuf_len = iov.iov_len;
+        return 0;
+}
+#endif
+
+/*
+ * Decompress a payload using the specified compression type. Allocates memory
+ * for uncompressed payload.
+ * @returns 0 on success, -1 on failure. Allocated memory in
+ * uncompressed_payload and its size in uncompressed_payload_size.
+ */
+int rd_kafka_telemetry_uncompress_metrics_payload(
+    rd_kafka_broker_t *rkb,
+    rd_kafka_compression_t compression_type,
+    void *compressed_payload,
+    size_t compressed_payload_size,
+    void **uncompressed_payload,
+    size_t *uncompressed_payload_size) {
+        int r = -1;
+        switch (compression_type) {
+#if WITH_ZLIB
+        case RD_KAFKA_COMPRESSION_GZIP:
+                *uncompressed_payload = rd_gz_decompress(
+                    compressed_payload, (int)compressed_payload_size,
+                    (uint64_t *)uncompressed_payload_size);
+                if (*uncompressed_payload == NULL)
+                        r = -1;
+                else
+                        r = 0;
+                break;
+#endif
+        case RD_KAFKA_COMPRESSION_LZ4:
+                r = rd_kafka_lz4_decompress(
+                    rkb, 0, 0, compressed_payload, compressed_payload_size,
+                    uncompressed_payload, uncompressed_payload_size);
+                break;
+#if WITH_ZSTD
+        case RD_KAFKA_COMPRESSION_ZSTD:
+                r = rd_kafka_zstd_decompress(
+                    rkb, compressed_payload, compressed_payload_size,
+                    uncompressed_payload, uncompressed_payload_size);
+                break;
+#endif
+#if WITH_SNAPPY
+        case RD_KAFKA_COMPRESSION_SNAPPY:
+                r = rd_kafka_snappy_decompress(
+                    rkb, compressed_payload, compressed_payload_size,
+                    uncompressed_payload, uncompressed_payload_size);
+                break;
+#endif
+        default:
+                fprintf(stderr, "Unknown compression type: %d\n",
+                        compression_type);
+                break;
+        }
+        return r;
+}
+
+/**
+ * Decode a metric from a buffer encoded with
+ * opentelemetry_proto_metrics_v1_MetricsData datatype. Used for testing and
+ * debugging.
+ *
+ * @param decode_interface The decode_interface to pass as arg when decoding the
+ * buffer.
+ * @param buffer The buffer to decode.
+ * @param size The size of the buffer.
+ */
+int rd_kafka_telemetry_decode_metrics(
+    rd_kafka_telemetry_decode_interface_t *decode_interface,
```

Comment:
Here too, either `interface` or `decode_interface` in other places too
```suggestion
    rd_kafka_telemetry_decode_interface_t *interface,
```

Reply:
I'd prefer to keep `decode_interface` everywhere